# Ido Weinstock - HW2 DS course

## Context

According to the World Health Organization (WHO) stroke is the 2nd leading cause of death globally, responsible for approximately 11% of total deaths.
This dataset is used to predict whether a patient is likely to get stroke based on the input parameters like gender, age, various diseases, and smoking status. Each row in the data provides relavant information about the patient.

## Attribute

The dataset used in this study is sourced from Kaggle (uploaded by fedesoriano), with its original medical origin designated as a confidential source. In healthcare data science, such confidentiality is standard practice rather than an anomaly, adhering to strict patient privacy regulations such as HIPAA and GDPR. Despite the absence of a specific institutional identifier, this dataset serves as a widely accepted benchmark in both academia and industry for addressing key machine learning challenges, including imbalanced classification and missing data imputation. Sanity checks and feature distributions—such as the clear correlations between age, hypertension, glucose levels, and stroke risk—demonstrate high consistency with established clinical literature, confirming the dataset's validity for exploratory data analysis and predictive modeling.

# EDA

## 1. Imports

*Dataset loading*

In [1]:
import pandas as pd
import numpy as np

import kagglehub

In [2]:
dataset_path = kagglehub.dataset_download("fedesoriano/stroke-prediction-dataset")
df = pd.read_csv(f"{dataset_path}/healthcare-dataset-stroke-data.csv")

df.head()

,id,gender,age,hypertension,heart_disease,ever_married,work_type,Residence_type,avg_glucose_level,bmi,smoking_status,stroke
0,9046,Male,67.0,0,1,Yes,Private,Urban,228.69,36.6,formerly smoked,1
1,51676,Female,61.0,0,0,Yes,Self-employed,Rural,202.21,NaN,never smoked,1
2,31112,Male,80.0,0,1,Yes,Private,Rural,105.92,32.5,never smoked,1
3,60182,Female,49.0,0,0,Yes,Private,Urban,171.23,34.4,smokes,1
4,1665,Female,79.0,1,0,Yes,Self-employed,Rural,174.12,24.0,never smoked,1


## 2. Data overview

### *Dataset shape*

In [3]:
df.shape

(5110, 12)

*Rows*: 5110

*Columns*: 12

*Variables*:
1) id: unique identifier
2) gender: "Male", "Female" or "Other"
3) age: age of the patient
4) hypertension: 0 if the patient doesn't have hypertension, 1 if the patient has hypertension
5) heart_disease: 0 if the patient doesn't have any heart diseases, 1 if the patient has a heart disease
6) ever_married: "No" or "Yes"
7) work_type: "children", "Govt_jov", "Never_worked", "Private" or "Self-employed"
8) Residence_type: "Rural" or "Urban"
9) avg_glucose_level: average glucose level in blood
10) bmi: body mass index
11) smoking_status: "formerly smoked", "never smoked", "smokes" or "Unknown"*
12) stroke: 1 if the patient had a stroke or 0 if not
*Note: "Unknown" in smoking_status means that the information is unavailable for this patient

In [4]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 5110 entries, 0 to 5109
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 5110 non-null   int64  
 1   gender             5110 non-null   str    
 2   age                5110 non-null   float64
 3   hypertension       5110 non-null   int64  
 4   heart_disease      5110 non-null   int64  
 5   ever_married       5110 non-null   str    
 6   work_type          5110 non-null   str    
 7   Residence_type     5110 non-null   str    
 8   avg_glucose_level  5110 non-null   float64
 9   bmi                4909 non-null   float64
 10  smoking_status     5110 non-null   str    
 11  stroke             5110 non-null   int64  
dtypes: float64(3), int64(4), str(5)
memory usage: 634.7 KB


### *Null values*

In [5]:
df.isnull().sum()

id                     0
gender                 0
age                    0
hypertension           0
heart_disease          0
ever_married           0
work_type              0
Residence_type         0
avg_glucose_level      0
bmi                  201
smoking_status         0
stroke                 0
dtype: int64

### *Dupliacations*

In [6]:
df.duplicated().sum()

np.int64(0)

### *Suspected values*

#### *negative values*

In [8]:
numeric_df = df.select_dtypes(include = 'number')
negative_values_df = numeric_df[numeric_df<0] # zero values won't be in that df

# .dropna(how='all'): Drops rows that contain only NaN values (rows without any negative numbers).
df_negatives_cleaned = negative_values_df.dropna(how="all")

print("--- DataFrame showing only negative values (non-negatives as NaN) ---")
df_negatives_cleaned.head()

--- DataFrame showing only negative values (non-negatives as NaN) ---


,id,age,hypertension,heart_disease,avg_glucose_level,bmi,stroke


There are no negative values

#### *age & marriage contradictions*

In [9]:
underage_marriage_df = df[
    (df["age"]< 18) & (df["ever_married"] == "Yes")
]

underage_marriage_df.head()

,id,gender,age,hypertension,heart_disease,ever_married,work_type,Residence_type,avg_glucose_level,bmi,smoking_status,stroke


There are no people under the leag age of marriage in the dataset

#### *BMI outliers*

In [12]:
bmi_outliers_df = df[
    (df["bmi"] < 10) & (df["bmi"]>60)
]

bmi_outliers_df.head()

,id,gender,age,hypertension,heart_disease,ever_married,work_type,Residence_type,avg_glucose_level,bmi,smoking_status,stroke


There are no BMI outliers